examples/go/tagging/handler/handler.go

```diff
@@ -31,140 +32,145 @@ type vcConfig struct {
 }
 
 // The structure for a Cloud Event.
-type vmEvent struct {
-	Event event `json:"data"`
-}
-type event struct {
-	Vm vmEventArgument `json:"Vm"`
-}
-type vmEventArgument struct {
-	Vm managedObjectReference `json:"Vm"`
-}
-type managedObjectReference struct {
-	Type  string `json:"Type"`
-	Value string `json:"Value"`
+type incoming struct {
+	Data types.Event `json:"data,omitempty"`
 }
 
+var (
+	client *vsClient.Client
+	ctx    = context.Background()
+	cfg    vcConfig
+	sigCh  = make(chan os.Signal)
+)
+
 // Handle a function invocation
 func Handle(req handler.Request) (handler.Response, error) {
-	ctx := context.Background()
-
-	// Retrieve vCenter and tag info from the config file.
-	cfg, err := loadTomlCfg()
-	if err != nil {
-		log.Error(err)
-		return
+	// Connect to vSphere govmomi API if not already done. Also, load the required
+	// vcconfig.
+	if client == nil {
+		err := vsConnect()
+		if err != nil {
+			log.Fatalf("connect to vSphere failed: %v", err)
+			return handler.Response{
+				Body:       nil,
+				StatusCode: http.StatusBadRequest,
+			}, fmt.Errorf("connect to vSphere failed: %w", err)
+		}
 	}
 
-	// Incoming request object.
-	var event vmEvent
-	err := json.Unmarshal(req.Body, &event)
+	// Set up os signal handling to log out of vSphere.
+	signal.Notify(sigCh, syscall.SIGTERM, os.Interrupt)
+	go handleSignal(ctx, sigCh)
+
+	// Retrieve the Managed Object Reference from the event.
+	moRef, err := parseEventMoRef(req.Body)
```

Comment:
Are we (handler.go) the server in this case? If so, then doesn't this part mean that we don't need to close anything?
> The Server will close the request body. The ServeHTTP Handler does not need to.

If we are client, then I'm very confused and my understanding of something is very wrong. But if we are client, yes, it does seem like we would be, as the docs say, "responsible for calling the Close method.for calling the Close method."

Reply:
We're "server" and likely net/http would help cleaning things up. BUT: it's considered good practice to always close io.ReadClosers after using them, e.g. via defer() instead of relying on someone else. Won't harm and trains muscle memory.

But again, the "body" we operate on is not of type io.ReadCloser anyways (I was confused by the variable names used)

Reply:
Because `body` is not type io.ReadCloser, I'll not close it. It is confusing that req has a field called Body.

```go
 type Response struct {

	// Body the body will be written back
	Body []byte

	// StatusCode needs to be populated with value such as http.StatusOK
	StatusCode int

	// Header is optional and contains any additional headers the function response should set
	Header http.Header
}
```
Source: https://github.com/openfaas-incubator/go-function-sdk/blob/master/handler.go